### PR TITLE
Feature/45 data validation

### PR DESCRIPTION
**Pull request by jamiebuckley**

### CODE REVIEW

File: app/models/contact.rb [hunk 1]
```diff
@@ -12,7 +12,14 @@ class Contact < ApplicationRecord
   has_paper_trail
 
   validates :first_name, presence: true
```
Comment (rik86future) [inline app/models/contact.rb:14]: I think these would be fairly easy to test in the contact_spec.rb

Comment (jamiebuckley) [reply app/models/contact.rb:14]: yeah good point, will add tests for those

File: app/models/contact.rb [hunk 2]
```diff
@@ -12,7 +12,14 @@ class Contact < ApplicationRecord
...
+  validates :surname, presence: true
   validates_date :date_of_birth, allow_nil: true, allow_blank: true
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, if: :email?
+  validates :telephone, numericality: true, length: { minimum: 9, maximum: 11 }, if: :telephone?
```
Comment (rik86future) [inline app/models/contact.rb:18]: the drawback is that we don't allow spaces any more for telephones

File: app/models/contact.rb [hunk 3]
```diff
@@ -12,7 +12,14 @@ class Contact < ApplicationRecord
   has_paper_trail
 
   validates :first_name, presence: true
+  validates :surname, presence: true
   validates_date :date_of_birth, allow_nil: true, allow_blank: true
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, if: :email?
```
Comment (rik86future) [inline app/models/contact.rb:17]: a*@gmail.com is not a valid email but this regex seems to accept it

File: app/models/contact.rb (hunk 2 above)
Comment (jamiebuckley) [reply app/models/contact.rb:18]: yes, that does introduce some complexities.  It's a little tricky validating length and also allowing spaces

File: app/models/contact.rb (hunk 3 above)
Comment (jamiebuckley) [reply app/models/contact.rb:17]: This regex comes from:
https://html.spec.whatwg.org/multipage/input.html#e-mail-state-(type=email)

* is a valid part of an email
https://en.wikipedia.org/wiki/Email_address#Local-part

Comment (rik86future) [reply app/models/contact.rb:17]: ok

**Comment by rik86future** [comment]
I think this PR is ok with the tests but I would be more comfortable to merge it if we also had the rake tasks to normalise existing data for prod...what do you think?

**Comment by jamiebuckley** [comment]
I agree, if you are happy with the PR then we'll just leave it open, and I'll start work on the data identification scripts.

**Pull request closed without merging**